ui/src/frontend/scroll_helper.ts

```diff
@@ -121,35 +121,60 @@ export function verticalScrollToTrack(
     return;
   }
 
-  let trackGroupId: string | undefined;
-  let trackSubgroupId: string | undefined;
-  let trackGroup = null;
-  let trackSubgroup = null;
+  // Track groups containing the track to be revealed, in order from top
+  // down, represented either as the rendered HTML element or, if not
+  // rendered because an ancestor group is collapsed, its UUID
+  const trackGroupsTopDown: (Element|string)[] = [];
   const containingIds = getContainingTrackIds(globals.state, trackIdString);
-  if (containingIds) {
-    trackGroupId = containingIds[0];
-    trackSubgroupId = containingIds[1];
-    trackGroup = document.querySelector('#track_' + trackGroupId);
-    if (trackSubgroupId) {
-      trackSubgroup = document.querySelector('#track_' + trackSubgroupId);
-    }
-  }
-
-  if (!trackGroupId || !trackGroup) {
+  if (!containingIds || !containingIds.length) {
     console.error(`Can't scroll, track (${trackIdString}) not found.`);
     return;
   }
 
-  // The requested track is inside a closed track group, either open the track
-  // group and scroll to the track or just scroll to the track group.
+  for (const trackGroupId of containingIds) {
+    const group = document.querySelector('#track_' + trackGroupId);
+
+    if (group) {
+      trackGroupsTopDown.push(group);
+    } else {
+      // Some containing group is collapsed, so this one is not rendered
+      trackGroupsTopDown.push(trackGroupId);
+    }
+  }
+
+  // The requested track is inside a closed track group. Either open the track
+  // group and scroll to the track or just scroll to deepest nested track group
+  // that has been rendered.
   if (openGroup) {
     // After the track exists in the dom, it will be scrolled to.
     globals.frontendLocalState.scrollToTrackId = trackId;
-    globals.dispatch(Actions.toggleTrackGroupCollapsed({trackGroupId}));
-    // TODO(cwd): Handle the track subgroup
+    for (let i = 1; i < trackGroupsTopDown.length; i++) {
+      const trackGroup = trackGroupsTopDown[i];
+      if (typeof trackGroup === 'string') {
+        // Expand its parent
+        if (typeof trackGroupsTopDown[i - 1] !== 'string') {
+          const trackGroupId = containingIds[i - 1];
+          globals.dispatch(Actions.toggleTrackGroupCollapsed({trackGroupId}));
+        }
+
+        // And mark all the rest for expansion when they are created
+        globals.frontendLocalState.expandTrackGroupIds.add(trackGroup);
+      } else if (i === (trackGroupsTopDown.length - 1)) {
+        // Need to expand the bottommost group to create the track to reveal
+        const trackGroupId = containingIds[i];
+        globals.dispatch(Actions.toggleTrackGroupCollapsed({trackGroupId}));
+      }
+    }
     return;
-  } else {
-    (trackSubgroup ?? trackGroup).scrollIntoView({behavior: 'smooth', block: 'nearest'});
+  }
+
+  // Find the deepest rendered group and scroll to it
```

Comment:
Do we not need to scroll to the target track (and not its group) if |openGroup| is false?

Also, a nit: could iterate without using an index in this case. Would look a bit cleaner. In the loops above, the index is referenced beyond just getting the collection element.

Reply:
@cdamus Could you please have a look at this?

Reply:
When `openGroup` is `false` the intent is that we do not expand groups as needed to reveal the track but instead scroll to reveal the containing group, which would remain collapsed. So extending this concept to the new support for arbitrary depth of grouping, we should scroll to reveal the _deepest_ containing group, which in this case is collapsed, and not expand it.

Reply:
Ah. My bad. I'm not sure how I misunderstood that `openGroup` argument so badly. It's described plainly and clearly in the comments. My brain must have shut down during the code review--I think it was getting late. That's my excuse. Also, ignore the nit comment. More nonsense.

Reply:
@jcortell68 too late! Commit 4dfeb3ea9d7b35a43a03240a860f71dec67256d4 addresses your review comments, so you're on the hook to re-review 😝